compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java

```diff
@@ -9486,16 +9502,8 @@ private STNode parseJoinClause(boolean isRhsExpr) {
         // allow-actions flag is always false, since there will not be any actions
         // within the from-clause, due to the precedence.
         STNode expression = parseExpression(OperatorPrecedence.QUERY, isRhsExpr, false);
+        STNode onCondition = parseOnClause(isRhsExpr);
         endContext();
```

Comment:
Shall we end the join-clause context before parsing the on-clause. like we had earlier?

* Since we already have a expr parsing(L9504) within join-clause context, cannot parse another expr under the same parent context as there would be an ambiguity. 
* In the error handler get next rule for the expression-rhs when the parent ctx is  `join-clause` includes on-clause as an alternative path. 
* Consider recovery for, `from int a in b join int d in e on <missing expr> select f`
after recoverying the missing expr, error handler will then check on-clause as the next rule. 
i.e could have two on-clauses after recovery

Reply:
Pleas update the doc comment above this method indicating optional on-clause as well

Reply:
According to the spec, on-clause is part of the join clause. So I assumed context should be closed ones everything is parsed. I'm confused here on what is `context` and how it works. Is there documentation which I can refer to?

Reply:
Spec says follows, and according to that `equals` is only valid within a join. So, If we close the join context before finishing up with the `join-on-condition`, how are we suppose to know, for example `on expression` should followed by an `equals`?

```
join-clause := "join" typed-binding-pattern "in" expression join-on-condition
join-on-condition := "on" expression "equals" expression
```